### PR TITLE
Add async/await support for watch expressions and callbacks

### LINKED ISSUE #102: Asyncio support
Support `async` watch expressions and callbacks!

### LINKED ISSUE #115: Initialize method
After #112 is merged, I want to add an `init` function to the public API that will automatically configure:

* asyncio eager task factory
* register the scheduler with asyncio and/or Qt

All depending on the versions of libs available and if they are already configured or not.

The function should also have an `async` version so it is compatible with `asyncio.run`. It probably just calls the synchronous function.

### PR DESCRIPTION
**Pull request by Korijn**
Closes #102
Closes #115 

* `async def` callbacks
* `async def` watch expressions (most notably for `watch_effect`)

True async (meaning `sync=False`) watch expressions (not callbacks) are only supported starting with python 3.12 with [`asyncio.eager_task_factory`](https://docs.python.org/3/library/asyncio-task.html?highlight=eager_task_factory#asyncio.eager_task_factory)

### CODE REVIEW

**Comment by Korijn** [comment]
There's a ton of differences between the way promises work in JS and in Python. For example, a promise will always execute eagerly to completion in JS, whereas in Python they will only complete if you explicitly await them. This makes the concept pretty difficult to implement... parking this for now.

**Comment by Korijn** [comment]
I have overlooked the `weak` function which wraps functions. That needs to support coroutines too.

**Comment by Korijn** [comment]
I think this is ready now!

File: tests/test_asyncio.py [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+        assert (called, completed) == (1, 1)
+    flush(loop)
+    if create_loop is create_plain_loop and expr_async and create_async:
+        assert (called, completed) == (0, 0)
+    elif (
+        create_loop is create_eager_loop and expr_async and create_async and write_async
+    ):
+        assert (called, completed) == (0, 0)
+    else:
+        assert (called, completed) == (1, 1)
```
Comment (Korijn) [inline tests/test_asyncio.py]: This section of the test is a little too much to handle for my brains. I just reverse engineered the `assert` statements from the outputs. I _think_ they make sense... but it's a lot to process.

Comment (ju1ius) [reply tests/test_asyncio.py]: IMO you should write one test function for each combination:
```py
def test_sync_watchable_with_sync_callback(): ...
def test_sync_watchable_with_async_callback(): ...

def test_async_watchable_with_sync_callback(): ...
def test_async_watchable_with_async_callback(): ...
```
Then maybe move the setup boilerplate to helper functions as needed.

At some point, and _especially_ in tests, some degree of duplication is better than combinatorial logic explosion that is hard to follow and refactor.

**Review by ju1ius — CHANGES_REQUESTED** [review]
@Korijn You should also update the `Watchable` type:

```py
Watchable = Union[
    Callable[[], T],
    Callable[[], Coroutine[Any, Any, T]],
    T,
]
```

File: tests/test_asyncio.py (hunk 1 above)
Comment (Korijn) [reply tests/test_asyncio.py]: Well, yes, but there are 48 cases in total... :(

In this case, it's pytest generating the cases. I can debug each individual case and check if it's acting correctly. I just need to go through all the cases one by one here and see if the applicable asserts make sense.

For `test_asyncio_watcheffect_multi`, the asserts are quite easy to read and they make sense to me. It confirms the implementation is correct.

For `test_asyncio_watcher_multi`, it's more complex. I just need a bit more time to dig in and convince myself it's correct.

Comment (Korijn) [reply tests/test_asyncio.py]: After reviewing more carefully, I think only the case `create_loop is create_eager_loop and expr_async and create_async and write_async` is suspicious and might be a bug. I'll step through that one.

Comment (Korijn) [reply tests/test_asyncio.py]: I figured it out, no bugs! I documented all the assert cases with extensive code comments.

**Comment by Korijn** [comment]
OK, for real this time, @berendkleinhaneveld, can you review this?

File: observ/watcher.py [hunk 2]
```diff
@@ -20,7 +21,12 @@
 
 
 T = TypeVar("T")
-Watchable = Union[Callable[[], T], T]
+Watchable = Union[
+    Callable[[], T],
+    Callable[[], Awaitable[T]],
+    T,
+    list[T],
```
Comment (ju1ius) [inline observ/watcher.py]: I'm not sure what you're trying to achieve here?

Comment (ju1ius) [reply observ/watcher.py]: This:
```py
a = proxy({'value': 'a'})
b = proxy({'value': 42})
w = watch([a, b], print, immediate=True, sync=True)
```
Should print
```py
[{'value': 'a'}, {'value': 42}]
```
Right?

If so, then you must _not_ include `list[T]` in the union.

Comment (ju1ius) [reply observ/watcher.py]: Let's test this:

```py
Watchable = Union[Callable[[], T], T]

a = proxy({'value': 'a'})
b = proxy({'value': 42})
w = watch([a, b], print, immediate=True, sync=True)
# [{'value': 'a'}, {'value': 42}]
typing.reveal_type(w) # Type of "w" is Watcher[list[dict[str, str] | dict[str: int]]]
```
Which is what we expect, but
```py
Watchable = Union[Callable[[], T], T, list[T]]

a = proxy({'value': 'a'})
b = proxy({'value': 42})
w = watch([a, b], print, immediate=True, sync=True)
# [{'value': 'a'}, {'value': 42}]
typing.reveal_type(w) # Type of "w" is Watcher[dict[str, str] | dict[str: int]]
```
is not.

Comment (ju1ius) [reply observ/watcher.py]:
```suggestion
```

**Review by ju1ius — CHANGES_REQUESTED** [review]

File: observ/watcher.py (hunk 2 above)
Comment (Korijn) [reply observ/watcher.py]: You're right, of course! That makes sense.

I really appreciate all the explanation and code review you're doing, I'm learning a lot from you. I'm not very experienced with type hints. Thanks!

**Review by ju1ius — APPROVED** [review]

File: observ/scheduler.py [hunk 3]
```diff
@@ -79,7 +79,7 @@ def register_qt(self):
             warnings.warn(
                 "QtAsyncio module available: please consider using `register_asyncio` "
                 "and call the following code:\n"
-                f"    from {qt} import QtAsyncio\n"
+                f"    from {qt} import QtAsyncio\n"  # noqa: E272
```
Comment (berendkleinhaneveld) [inline observ/scheduler.py:81]: Was this put here intentionally?

**Review by berendkleinhaneveld — APPROVED** [review]
I have read most of the comments, but its a bit much for me, parsing them and really understanding them within a limited timeframe. I say: LGTM: I think it looks great! And I think it is in a great state should we encounter problems later on and have to dig in again.

File: observ/scheduler.py (hunk 3 above)
Comment (Korijn) [reply observ/scheduler.py:81]: Yes. It only triggers when linting on python 3.12 :/ I ran into that while working locally
